### PR TITLE
add model inversion post

### PR DESCRIPTION
**Pull request by skeydan**

### CODE REVIEW

File: _posts/2020-05-15-model-inversion-attacks/model_inversion_differential_privacy.Rmd [hunk 1]
```diff
@@ -0,0 +1,740 @@
...
+  
+}
+
+model <- target_model()
+
+```
+
+The overall model is a Keras custom model, so we train it [TensorFlow 2.x -
+style](https://tensorflow.rstudio.com/tutorials/advanced/). After ten epochs, training and validation accuracy are at
+~0.84\ and~ 0.73, respectively -- not bad at all for a 20-class discrimination task.
```
Comment (javierluraschi) [inline _posts/2020-05-15-model-inversion-attacks/model_inversion_differential_privacy.Rmd:227]: This seems to be causing some weird rendering, intended?

File: _posts/2020-05-15-model-inversion-attacks/model_inversion_differential_privacy.Rmd [hunk 2]
```diff
@@ -0,0 +1,740 @@
+---
+title: "Hacking deep learning: model inversion attack by example"
+description: > 
+  Compared to other applications, deep learning models might not seem too likely as victims of privacy attacks. However, methods exist to determine whether an entity was used in the training set (an adversarial attack called member inference), and techniques subsumed under "model inversion" allow to spy on raw data input given just model output (and sometimes, context information). This post shows an end-to-end example of model inversion, and explores mitigation strategies using TensorFlow Privacy.
```
Comment (batpigandme) [inline _posts/2020-05-15-model-inversion-attacks/model_inversion_differential_privacy.Rmd:4]: > allow to spy on raw data input given just model output (and sometimes, context information)

I think you need a word between "allow" and "to spy" (allow _what_ to spy?)

File: _posts/2020-05-15-model-inversion-attacks/model_inversion_differential_privacy.Rmd [hunk 3]
```diff
@@ -0,0 +1,740 @@
...
+preview: images/results.png
+---
+
+```{r setup, include=FALSE}
+knitr::opts_chunk$set(echo = TRUE, eval = FALSE)
+```
+
+
+How private are individual data in the context of machine learning models? The data used to train the model, say. There are
+types of models where the answer is simple. Take k-nearest-neighbors, for example. There *is* not even a model without the
```
Comment (batpigandme) [inline _posts/2020-05-15-model-inversion-attacks/model_inversion_differential_privacy.Rmd:28]:
```suggestion
types of models where the answer is simple. Take k-nearest-neighbors, for example. There *is not* even a model without the
```

File: _posts/2020-05-15-model-inversion-attacks/model_inversion_differential_privacy.Rmd [hunk 4]
```diff
@@ -0,0 +1,740 @@
...
+complete dataset. Or support vector machines. There is no model without the support vectors. But neural networks? They're just
+some composition of functions, -- no data included.
+
+Same with data fed to a deployed deep learning model. It's pretty unlikely one could invert the final softmax output from a
```
Comment (batpigandme) [inline _posts/2020-05-15-model-inversion-attacks/model_inversion_differential_privacy.Rmd:32]:
```suggestion
The same is true for data fed to a deployed deep-learning model. It's pretty unlikely one could invert the final softmax output from a
```

File: _posts/2020-05-15-model-inversion-attacks/model_inversion_differential_privacy.Rmd [hunk 5]
```diff
@@ -0,0 +1,740 @@
...
+big ResNet and get back the raw input data.
+
+In theory, then, "hacking" a standard neural net to spy on input data sounds illusory. In practice, however, there is always
+some real-world *context*. The context may be other datasets, publicly available, that can be linked to the "private" data in
+question. This is a popular showcase used in advocating for differential privacy[@Dwork2006]: Take an "anonymized" dataset,
+dig up complementary information from public sources, and de-anonymize records ad libitum. Some context in that sense will
+often be used in "black-box" attacks, ones that presuppose no insider information about the model to be hacked.
+
+But context can also be structural, such as in the scenario demonstrated in this post. For example, assume a distributed
+model, where sets lof layers run on different devices (embedded devices or mobile phones, for example). (A scenario like that
```
Comment (batpigandme) [inline _posts/2020-05-15-model-inversion-attacks/model_inversion_differential_privacy.Rmd:42]:
```suggestion
model, where sets of layers run on different devices (embedded devices or mobile phones, for example). (A scenario like that
```

Comment (batpigandme) [reply _posts/2020-05-15-model-inversion-attacks/model_inversion_differential_privacy.Rmd:42]: Having a parenthetical right after a sentence that ends with a parenthetical just feels off to me.

File: _posts/2020-05-15-model-inversion-attacks/model_inversion_differential_privacy.Rmd [hunk 6]
```diff
@@ -0,0 +1,740 @@
...
+is sometimes seen as "white-box"[@7536387], but in common understanding, white-box attacks probably presuppose some more
+insider knowledge, such as access to model architecture or even, weights. I'd therefore prefer calling this white-ish at
+most.) --- Now assume that in this context, it is possible to intercept, and interact with, a system that executes the deeper
+layers of the model. Based on that system's intermediate-level output, it is possible to perform *model inversion*[@Fred],
+that is, reconstruct the input data fed into the system.
```
Comment (batpigandme) [inline _posts/2020-05-15-model-inversion-attacks/model_inversion_differential_privacy.Rmd:47]:
```suggestion
that is, to reconstruct the input data fed into the system.
```

File: _posts/2020-05-15-model-inversion-attacks/model_inversion_differential_privacy.Rmd [hunk 7]
```diff
@@ -0,0 +1,740 @@
...
+
+In this post, we'll demonstrate such a model inversion attack, basically porting the approach given in a
```
Comment (batpigandme) [inline _posts/2020-05-15-model-inversion-attacks/model_inversion_differential_privacy.Rmd:49]:
```suggestion
In this post, we'll demonstrate such a model-inversion attack, basically porting the approach given in a
```

File: _posts/2020-05-15-model-inversion-attacks/model_inversion_differential_privacy.Rmd [hunk 8]
```diff
@@ -0,0 +1,740 @@
...
+[notebook](https://github.com/OpenMined/PySyft/blob/master/examples/tutorials/advanced/privacy_attacks/Tutorial%201%20-%20Black%20box%20model%20inversion.ipynb Sy)
+found in the [PySyft](https://github.com/OpenMined/PySyft) repository. We then experiment with different levels of
+$\epsilon$-privacy, exploring impact on reconstruction success. This second part will make use of TensorFlow Privacy,
```
Comment (batpigandme) [inline _posts/2020-05-15-model-inversion-attacks/model_inversion_differential_privacy.Rmd:52]: Add link to TensorFlow Privacy here for first mention?

File: _posts/2020-05-15-model-inversion-attacks/model_inversion_differential_privacy.Rmd [hunk 9]
```diff
@@ -0,0 +1,740 @@
...
+
+## Part 1: Model inversion in action
+
+### Example dataset: All the world's letters[^1]
+
+[^1]: Don't take *all* literally please; it's just a nice phrase.
+
+The overall process of model inversion used here is the following. With no, or scarcely any, insider knowledge about a model,
+-- but given opportunities to repeatedly query it --, I want to learn how to reconstruct unknown inputs based on just model
+outputs . Independently of original model training, this, too, is a training process; however, in general it will not involve
```
Comment (batpigandme) [inline _posts/2020-05-15-model-inversion-attacks/model_inversion_differential_privacy.Rmd:63]:
```suggestion
outputs . Independent of original model training, this, too, is a training process; however, in general it will not involve
```

File: _posts/2020-05-15-model-inversion-attacks/model_inversion_differential_privacy.Rmd [hunk 10]
```diff
@@ -0,0 +1,740 @@
...
+the original data, as those won't be publicly available. Still, for best success, the attacker model is trained with data as
+similar as possible to the original training data assumed. Thinking of images, for example, and presupposing the popular view
+of successive layers representing successively coarse-grained features, we want that the surrogate data to share as many
```
Comment (batpigandme) [inline _posts/2020-05-15-model-inversion-attacks/model_inversion_differential_privacy.Rmd:66]:
```suggestion
of successive layers representing successively coarse-grained features, we want the surrogate data to share as many
```

File: _posts/2020-05-15-model-inversion-attacks/model_inversion_differential_privacy.Rmd [hunk 11]
```diff
@@ -0,0 +1,740 @@
...
+representation spaces with the real data as possible -- up to the very highest layers before final classification, ideally.
+
+If we wanted to use classical MNIST as an example, one thing we could do is
+to only use some of the digits for training the "real" model; and the rest, for training the adversary. Let's try something
```
Comment (batpigandme) [inline _posts/2020-05-15-model-inversion-attacks/model_inversion_differential_privacy.Rmd:70]:
```suggestion
to use only some of the digits for training the "real" model; and the rest, for training the adversary. Let's try something
```

File: _posts/2020-05-15-model-inversion-attacks/model_inversion_differential_privacy.Rmd [hunk 12]
```diff
@@ -0,0 +1,740 @@
...
+different though, something that might make the undertaking harder as well as easier at the same time. Harder, because the
+dataset features exemplars more complex than MNIST digits; easier because of the same reason: More could possibly be learned,
+by the adversary, from a complex task.
```
Comment (batpigandme) [inline _posts/2020-05-15-model-inversion-attacks/model_inversion_differential_privacy.Rmd:73]:
```suggestion
dataset features exemplars more complex than MNIST digits; easier because of the same reason: More could possibly be learned
by the adversary from a complex task.
```

File: _posts/2020-05-15-model-inversion-attacks/model_inversion_differential_privacy.Rmd [hunk 13]
```diff
@@ -0,0 +1,740 @@
...
+sample:
+
+```{r, eval=TRUE, echo=FALSE, fig.cap = "Sample from the twenty-alphabet set used to train the target model (originally: 'evaluation set')"}
+knitr::include_graphics("images/training_set_images.png")
+```
+
+The group of thirty we don't use; instead, we'll employ two small five-alphabet collections to train the adversary and to test
+reconstruction, respectively. (These small subsets of the original "big" thirty-alphabet set are again disjoint.)
+
+Here first is a sample from the set used to train the adversary.
```
Comment (batpigandme) [inline _posts/2020-05-15-model-inversion-attacks/model_inversion_differential_privacy.Rmd:87]:
```suggestion
Here, first, is a sample from the set used to train the adversary.
```

File: _posts/2020-05-15-model-inversion-attacks/model_inversion_differential_privacy.Rmd [hunk 14]
```diff
@@ -0,0 +1,740 @@
...
+Now first, we train the target model.
+
+### Train target model
+
+The dataset originally has four columns: the image, of size 105 x 105; an alphabet id and a within-dataset character id; and a
+label. For our use case, we're not really interested in the task the target model was/is used for; we just want to get at the
+data. Basically, whatever task we choose, it is not much more than a dummy task. So, let's just say we train the target to
+classify characters *by alphabet*.
+
+We thus throw out all unneeded features, keeping just the alphabet id and the image itself:
```
Comment (batpigandme) [inline _posts/2020-05-15-model-inversion-attacks/model_inversion_differential_privacy.Rmd:133]:
```suggestion
Thus, we throw out all unneeded features, keeping just the alphabet id and the image itself:
```

File: _posts/2020-05-15-model-inversion-attacks/model_inversion_differential_privacy.Rmd [hunk 15]
```diff
@@ -0,0 +1,740 @@
...
+    record$image <- preprocess_image(record$image)
+    list(record$image, record$alphabet)}) %>%
+  dataset_batch(32)
+```
+
+The model consists of two parts. The first is imagined to run in a distributed fashion; for example, on mobile devices (stage
+one). These devices then send model outputs to a central server, where final results are computed (stage two). Sure, you will
+be thinking, this is a convenient setup for our scenario: If we intercept stage one results, we -- most probably -- gain
+access to richer information than what is contained in a model's final output layer. --- That is correct, but the scenario is
+less contrived than one might assume. Just like federated learning [@McMahanMRA16], it fulfills important desiderata: Actual
```
Comment (batpigandme) [inline _posts/2020-05-15-model-inversion-attacks/model_inversion_differential_privacy.Rmd:166]: I love the word desiderata (I literally have a GitHub repo by that name), but I think it's a bit esoteric in this context. 🤓

File: _posts/2020-05-15-model-inversion-attacks/model_inversion_differential_privacy.Rmd [hunk 16]
```diff
@@ -0,0 +1,740 @@
...
+training data never leaves the devices, thus staying (in theory!) private; at the same time, ingoing traffic to the server is
+significantly reduced.
+
+In our example setup, the on-device model is a convnet, while the server model is a simple feedforward network.
+
+We link both together as a *TargetModel* that when called normally, will run both steps in succession. However, we'll be able
+to call `target_model$mobile_step()` separately, thus intercepting intermediate results.
```
Comment (batpigandme) [inline _posts/2020-05-15-model-inversion-attacks/model_inversion_differential_privacy.Rmd:173]:
```suggestion
to call `target_model$mobile_step()` separately, thereby intercepting intermediate results.
```

Comment (batpigandme) [reply _posts/2020-05-15-model-inversion-attacks/model_inversion_differential_privacy.Rmd:173]: Just mixing the words up a bit for kicks.

File: _posts/2020-05-15-model-inversion-attacks/model_inversion_differential_privacy.Rmd [hunk 17]
```diff
@@ -0,0 +1,740 @@
...
+### Train adversary
+
+The adversary's general strategy will be:
+
+-   Feed its small, surrogate dataset to the on-device model. The output received can be regarded as a (highly) *compressed*
+    version of the original images.
+-   Pass that "compressed" version as input to its own model, which tries to reconstruct the original images from the sparse
+    code.
+-   Compare original images (those from the surrogate dataset) to the reconstruction pixel-wise. The goal is to minimize the
+    mean (squared, say) error.
```
Comment (batpigandme) [inline _posts/2020-05-15-model-inversion-attacks/model_inversion_differential_privacy.Rmd:306]: I would embolden the action sentence in each of these, e.g., **Feed its small, surrogate dataset to the on-device model.**, or something to that effect, to make the list a bit more skimmable.

File: _posts/2020-05-15-model-inversion-attacks/model_inversion_differential_privacy.Rmd [hunk 18]
```diff
@@ -0,0 +1,740 @@
...
+  dataset_shuffle(1000) %>%
+  # need dataset_repeat() when not eager
+  dataset_repeat() %>%
+  dataset_batch(batch_size)
+```
+
+### Train target model -- with TensorFlow Privacy
+
+To train the target, we put the layers from both stages -- "mobile" and "server" -- into one sequential model. Note how we
+remove the dropout. This is because anyway, noise will be added during optimization.
```
Comment (batpigandme) [inline _posts/2020-05-15-model-inversion-attacks/model_inversion_differential_privacy.Rmd:546]:
```suggestion
remove the dropout. This is because noise will be added during optimization anyway.
```

File: _posts/2020-05-15-model-inversion-attacks/model_inversion_differential_privacy.Rmd [hunk 19]
```diff
@@ -0,0 +1,740 @@
...
+```
+
+And now, we're back to the old training routine. Testing setup is the same as before, as well.
+
+So how well does the adversary perform with differential privacy added to the picture?
+
+### Test adversary (against differentially private target)
+
+Here, ordered by decreasing $\epsilon$, are the reconstructions. Again, we refrain from judging the results, for the same
+reasons as before: In real-world applications, whether privacy is preserved well "enough" will depend on the context.
```
Comment (batpigandme) [inline _posts/2020-05-15-model-inversion-attacks/model_inversion_differential_privacy.Rmd:707]:
```suggestion
reasons as before: In real-world applications, whether privacy is preserved "well enough" will depend on the context.
```

File: _posts/2020-05-15-model-inversion-attacks/model_inversion_differential_privacy.Rmd [hunk 20]
```diff
@@ -0,0 +1,740 @@
...
+Throughout this post, we've refrained from "over-commenting" on results, and focused on the why-and-how instead. This is
+because in an artificial setup, chosen to facilitate exposition of concepts and methods, there really is no objective frame of
+reference. What is a good reconstruction? What is a good $\epsilon$? What constitutes a data breach? No-one knows.
+
+In the real world, there is a context to everything -- there are people involved, the people whose data we're talking about.
+There are organizations, regulations, laws. There are abstract principles, and there are implementations; different
+implementations of the same "idea" can differ.
+
+As in machine learning overall, research papers on privacy-, ethics- or otherwise society-related topics are full of LaTeX
+formulae. Between the math, let's not forget the people.
```
Comment (batpigandme) [inline _posts/2020-05-15-model-inversion-attacks/model_inversion_differential_privacy.Rmd:738]:
```suggestion
formulae. Amid the math, let's not forget the people.
```

File: _posts/2020-05-15-model-inversion-attacks/model_inversion_differential_privacy.Rmd (hunk 2 above)
Comment (skeydan) [reply _posts/2020-05-15-model-inversion-attacks/model_inversion_differential_privacy.Rmd:4]: hm... or another verb...? reconstruct, I think...

File: _posts/2020-05-15-model-inversion-attacks/model_inversion_differential_privacy.Rmd (hunk 16 above)
Comment (skeydan) [reply _posts/2020-05-15-model-inversion-attacks/model_inversion_differential_privacy.Rmd:173]: right ... :-) i should consult the synonyms site more often ...

File: _posts/2020-05-15-model-inversion-attacks/model_inversion_differential_privacy.Rmd (hunk 18 above)
Comment (skeydan) [reply _posts/2020-05-15-model-inversion-attacks/model_inversion_differential_privacy.Rmd:546]: hehe, rhetorical "syntax inversion" again ;-)

File: _posts/2020-05-15-model-inversion-attacks/model_inversion_differential_privacy.Rmd (hunk 20 above)
Comment (skeydan) [reply _posts/2020-05-15-model-inversion-attacks/model_inversion_differential_privacy.Rmd:738]: oooh right...

File: _posts/2020-05-15-model-inversion-attacks/model_inversion_differential_privacy.Rmd (hunk 17 above)
Comment (skeydan) [reply _posts/2020-05-15-model-inversion-attacks/model_inversion_differential_privacy.Rmd:306]: makes sense!

File: _posts/2020-05-15-model-inversion-attacks/model_inversion_differential_privacy.Rmd (hunk 1 above)
Comment (skeydan) [reply _posts/2020-05-15-model-inversion-attacks/model_inversion_differential_privacy.Rmd:227]: right, thanks! Will correct...

**Comment by skeydan** [comment]
thanks a ton Mara!!
